components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt

```diff
@@ -67,3 +68,19 @@ fun String.sha1(): String {
         String(charArrayOf(characters[byte.toInt() shr 4 and 0x0f], characters[byte.toInt() and 0x0f]))
     })
 }
+
+/**
+ * Converts String to date, trying multiple formats
+ */
+fun String.getDateFromString(): Date? {
+
+    val possibleDateFormats = arrayOf("yyyy-MM-dd'T'HH:mm", "yyyy-MM-dd", "yyyy-'W'ww", "yyyy-MM", "HH:mm")
```

Comment:
I'm not quite sure if trying to guess the format could be the right approach, because this could lead us to unexpected behaviors, in some cases, we may be doing the right thing and others not. As the date was not in the correct format since the beginning, we really don't know what was the initial intention of the developer we may pick a wrong format trying to recovering. 

This is hard question to answer, I checked for some references (they are different between each other):

**Android:**
Fennec and Chrome do not try to guess the format allowing the user to pick whatever date.
The site is unusable if you don't pick a date between today and a month after, the dates are never picked.

**Desktop:**
FireFox and Chrome they try to guess the format allowing the user to pick dates from `minDate` as today  until `maxDate` a month after today.

Reply:
🤔  I'm not sure which approach we follow here desktop or mobile(Android)

Reply:
IMO, on Android, Chrome and Fennec don't do a very good job, ignoring minDate and maxDate, letting user select invalid dates.

On desktop, making minDate today and maxDate random în the future ignores the website's developer constraints.

We can also limit the formats we can expect, by removing the ones that could generate errors from the list, but this way we can account for the values send by the website.

Reply:
In the guessing we are also doing wrong for instance the `min` and `max` date they look like `yyyy-MM-dd` but the real developer intention was `yyyy-MM-dd'T'HH:mm`. On Fennec and chrome for Android try to select today but with a time before `16:16`(04:16PM)

"2019-08-16 16:16" -> ✅  The date is set
**Clear the input**
"2019-08-16 16:15" -> 🔴   The date is not set

Reply:
Yes, that is why I put the format with time first in the list. The main usage for the min limit is to prevent selecting a time before now() on searches for booking, routes, etc, cause that would render them obsolete.

Reply:
Yes, but what we are getting from the website as `minDate` is `2019-08-16` and that will fall under `yyyy-MM-dd` allowing users to select any time before `16:16` that was the developer intention as we can see because they are clearing out the input.

This is why I'm not completely sure if guessing is the right approach because. In guessing, we are not completely sure if we are doing the right thing.

Before leaning to one approach or another I would like to hear the third opinion @pocmo what do you think?

Reply:
@mcarare  I'm trying to find what FireFox desktop is doing because they are guessing the full date including the time. At least this way we are going to have an aligned behavior across platforms, otherwise we can use the approach that you suggested above 👍

Reply:
I couldn't find what exactly desktop ([The farder that I got was](https://dxr.mozilla.org/mozilla-central/source/dom/html/HTMLInputElement.cpp#1721)) was doing to guess the date format. But I think your approach it is good alternative :) 👍 

One last, suggestion: I think `String.getDateFromString()` could be improved if we rename to
`String.toDate` that will allow the consumers to change the `possibleFormats`.

```kotlin
/**
 * Tries to convert a [String] to a [Date] using a list of [possibleFormats].
 * @param possibleFormats one ore more possible format.
 * @return a [Date] object with the values in the provided in this string,
 * if the conversion is not possible null will be returned.
 */
fun String.toDate(
    vararg possibleFormats: String = arrayOf(
        "yyyy-MM-dd'T'HH:mm",
        "yyyy-MM-dd",
        "yyyy-'W'ww",
        "yyyy-MM",
        "HH:mm"
    )
): Date? {
    possibleFormats.forEach {
        try {
            return this.toDate(it)
        } catch (pe: ParseException) {
            // move to next possible format
        }
    }
    return null
}
```

What do you think?

Reply:
I think this is a good approach! Will you make these changes, or should I?

Reply:
I can do the change :) 
Thanks!